src/test.js

```diff
@@ -466,20 +461,43 @@ function generateHash( module, testName ) {
 	return hex.slice( -8 );
 }
 
-function synchronize( callback, last ) {
+function synchronize( callback, priority ) {
+	var last = !priority;
+
 	if ( QUnit.objectType( callback ) === "array" ) {
 		while ( callback.length ) {
 			synchronize( callback.shift() );
 		}
 		return;
 	}
-	config.queue.push( callback );
+
+	if ( priority ) {
+		priorityFill( callback );
+	} else {
+		config.queue.push( callback );
+	}
 
 	if ( config.autorun && !config.blocking ) {
 		process( last );
 	}
 }
 
+// Place previously failed tests on a queue priority line, respecting the order they get assigned.
+function priorityFill( callback ) {
+	var queue, prioritizedQueue;
+
+	queue = config.queue.slice( priorityFill.pos );
+	prioritizedQueue = config.queue.slice( 0, -config.queue.length + priorityFill.pos );
+
+	queue.unshift( callback );
+	queue.unshift.apply( queue, prioritizedQueue );
+
+	config.queue = queue;
+
+	return priorityFill.pos++;
```

Comment:
As you noticed, this return does not improve anything, I'll remove it. Even though, I'm getting paranoid to add `return` on every function because of Tail Call Optimization, spec'ed on ES2015 and which will affect JS engines silently.

The TCO avoids memory leaks and brings a slight performance optimization when we use it.

With that said, it **only** affects function returns of other function calls (the tail call), like `return anything();`. The subroutine function will reuse the current stack frame (the scope stack) without creating a new one.

I know I explained more than the necessary, I just wanted to share this. It is awesome, right? We should make TCO calls a recommendation on our style guide.